sources/pam.c

```diff
@@ -0,0 +1,88 @@
+/*
+ * Odyssey.
+ *
+ * Scalable PostgreSQL connection pooler.
+*/
+
+#include <security/pam_appl.h>
+#include <kiwi.h>
+
+struct auth_request {
+    char *password;
+};
+
+
+static int
+pam_conversation(int msgc,
+		const struct pam_message **msgv,
+	       	struct pam_response **rspv,
+		void *authdata) {
+	struct auth_request *request = (struct auth_request *) authdata;
+	int counter, retcode;
+	
+	if (msgc < 1 || msgv == NULL || request == NULL) {
+		return PAM_CONV_ERR;
+	}
+	*rspv = malloc(msgc * sizeof(struct pam_response));
+	if (*rspv == NULL) {
+		return PAM_CONV_ERR;
+	}
+	
+	memset(*rspv, 0, msgc * sizeof(struct pam_response));
+	retcode = PAM_SUCCESS;
+	
+	for (counter = 0; retcode == PAM_SUCCESS && counter < msgc; counter++) {
+		if (msgv[counter]->msg_style == PAM_PROMPT_ECHO_OFF) {
+			(*rspv)[counter].resp = strdup(request->password);
+			if ((*rspv)[counter].resp == NULL) {
+				retcode = PAM_CONV_ERR;
```

Comment:
should we call `break;` here as well?

Reply:
There was sufficient loop condition for that (retcode == PAM_SUCCESS && counter < msgc), but ok, maybe it was bad for readability, so i add if statement to be more explicitly